File: ch2/lineards/list.ml
(* This source code is not as complete as list.cpp *)

open Printf

let () =
  let s = Stack.create () in
  let q = Queue.create () in

  printf "%B\n" (Stack.is_empty s);       (* currently s is empty, true *)
  printf "==================\n";
  Stack.push 'a' s;
  Stack.push 'b' s;
  Stack.push 'c' s;
  (* stack is LIFO, thus the content of s is currently like this:
   * c <- top
   * b
   * a *)
  printf "%c\n" (Stack.top s);                            (* output 'c' *)
  Stack.pop s |> ignore;                                 (* pop topmost *)
  printf "%c\n" (Stack.top s);                            (* output 'b' *)
  printf "%B\n" (Stack.is_empty s);  (* currently s is not empty, false *)
  printf "==================\n";

  printf "%B\n" (Queue.is_empty q);       (* currently q is empty, true *)
  printf "==================\n";
  while (not (Stack.is_empty s)) do   (* stack s still has 2 more items *)
    Queue.push (Stack.pop s) q;            (* enqueue 'b', and then 'a' *)
  done;
  Queue.push 'z' q;                                (* add one more item *)
  printf "%c\n" (Queue.top q);                            (* prints 'b' *)

  (* output 'b', 'a', then 'z' (until queue is empty), according to the insertion order above *)
  printf "==================\n";
  while (not (Queue.is_empty q)) do
    printf "%c\n" (Queue.peek q);               (* take the front first *)
    Queue.pop q |> ignore            (* before popping (dequeue-ing) it *)
  done;

File: ch2/lineards/list.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  stack<char> s;
  printf("%d\n", s.empty());                   // currently s is empty, true (1)
  printf("==================\n");
  s.push('a');
  s.push('b');
  s.push('c');
  // stack is LIFO, thus the content of s is currently like this:
  // c <- top
  // b
  // a
  printf("%c\n", s.top());                       // output 'c'
  s.pop();                                       // pop topmost
  printf("%c\n", s.top());                       // output 'b'
  printf("%d\n", s.empty());                     // output false (0)
  printf("==================\n");

  queue<char> q;
  printf("%d\n", q.empty());                     // output true (1)
  printf("==================\n");
  while (!s.empty()) {                           // s has 2 more items
    q.push(s.top());                             // enqueue 'b' and 'a'
    s.pop();
  }
  q.push('z');                                   // add one more item
  printf("%c\n", q.front());                     // prints 'b'
  printf("%c\n", q.back());                      // prints 'z'

  // output 'b', 'a', then 'z' (until queue is empty), according to the insertion order above
  printf("==================\n");
  while (!q.empty()) {
    printf("%c\n", q.front());                   // take the front first
    q.pop();                                     // before popping it
  }

  printf("==================\n");
  deque<char> d;
  d.push_back('a');
  d.push_back('b');
  d.push_back('c');
  printf("%c - %c\n", d.front(), d.back());      // prints 'a - c'
  d.push_front('d');
  printf("%c%c%c%c\n", d[0], d[1], d[2], d[3]);  // only in C++, 'dabc'
  printf("%c - %c\n", d.front(), d.back());      // prints 'd - c'
  d.pop_back();
  printf("%c - %c\n", d.front(), d.back());      // prints 'd - b'
  d.pop_front();
  printf("%c - %c\n", d.front(), d.back());      // prints 'a - b'

  return 0;
}
